service/storage_service.cc

```diff
@@ -998,6 +998,20 @@ class topology_coordinator {
         }
     }
 
+    future<std::pair<node_to_work_on, bool>> global_token_metadata_barrier(node_to_work_on&& node) {
+        bool exec_command_res;
+        std::tie(node, exec_command_res) = co_await exec_global_command(std::move(node),
+            raft_topology_cmd { raft_topology_cmd::command::barrier_and_drain },
+            true);
+        if (!exec_command_res) {
+            co_return std::pair{std::move(node), exec_command_res};
```

Comment:
I don't like exceptions in general since they complicate control flow. The additional channel for returning results from functions complicate matters, it's difficult to reason about it, whether we covered all the potential exception types in the corresponding catch on not.

In this example, the only difference will be `if` replaced with `try/catch`. I would like to hear from @gleb-cloudius, since he is the author of this code.

Reply:
> Not introduced by this patch, but `exec_global_command_helper()` should not return a bool, but an exception. Returning bool will make debugging issues much harder, since you have to dig the error from the logs on a different node, and you won't know which error belongs to which failure.

I do not understand how throwing an exception will make any difference. And as we learned exceptions should not be used for flow control.

Reply:
> In this example, the only difference will be if replaced with try/catch. I would like to hear from @gleb-cloudius, since he is the author of this code.

You can keep the `if`, just check `f.failed()` and throw or log `f.get_exception()`.

> I do not understand how throwing an exception will make any difference. 

The caller of global_token_metadata_barrier() will get the cause of exec_global_command() failure, so instead of logging "global barrier failed" it will log "global barrier failed: node 127.0.0.2: std::bad_alloc()"

> And as we learned exceptions should not be used for flow control.

Because of performance?

Reply:
> > In this example, the only difference will be if replaced with try/catch. I would like to hear from @gleb-cloudius, since he is the author of this code.
> 
> You can keep the `if`, just check `f.failed()` and throw or log `f.get_exception()`.

That what `exec_global_command_helper` does exactly, no?


> 
> > I do not understand how throwing an exception will make any difference.
> 
> The caller of global_token_metadata_barrier() will get the cause of exec_global_command() failure, so instead of logging "global barrier failed" it will log "global barrier failed: node 127.0.0.2: std::bad_alloc()"

OK, but the error is already logged by whoever returned the failure.

> 
> > And as we learned exceptions should not be used for flow control.
> 
> Because of performance?

Yes.

Reply:
> > > In this example, the only difference will be if replaced with try/catch. I would like to hear from @gleb-cloudius, since he is the author of this code.
> > 
> > 
> > You can keep the `if`, just check `f.failed()` and throw or log `f.get_exception()`.
> 
> That what `exec_global_command_helper` does exactly, no?

Yes. The point was that we don't need to add try/catch when we work with exceptions, which was brought up by @gusev-p.

> 
> > > I do not understand how throwing an exception will make any difference.
> > 
> > 
> > The caller of global_token_metadata_barrier() will get the cause of exec_global_command() failure, so instead of logging "global barrier failed" it will log "global barrier failed: node 127.0.0.2: std::bad_alloc()"
> 
> OK, but the error is already logged by whoever returned the failure.

So the information is probably there. But if you were to compose the failure report in a hierarchy, and log that a higher operation failed, it makes it hard to analyze the failure if the cause is in a different line.  Also, information about causality may be lost this way. For example, are you sure that the failure you see on the cmd level belongs is the failure on a higher level, or maybe it belongs to an earlier retry which eventually succeed. I can see there is no such logging now, but we probably will want it at some point.

> 
> > > And as we learned exceptions should not be used for flow control.
> > 
> > 
> > Because of performance?
> 
> Yes.

In this case, it should be rare. We should add sleep between failure and retry so that we don't retry+fail in a loop which is too tight.

Reply:
> So the information is probably there. But if you were to compose the failure report in a hierarchy, and log that a higher operation failed, it makes it hard to analyze the failure if the cause is in a different line. Also, information about causality may be lost this way. For example, are you sure that the failure you see on the cmd helper level is the cause of a failure on a higher level, or maybe it belongs to an earlier retry which eventually succeed. I can see there is no such logging now, but we probably will want it at some point.

I added [a commit](https://github.com/scylladb/scylladb/pull/13884/commits/b4d0025e10c1dd1df22d83a3a45202eef85b571a) with the refactoring.

> In this case, it should be rare. We should add sleep between failure and retry so that we don't retry+fail in a loop which is too tight.

Let's postpone this until the proper error handling implementation.